presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -666,9 +667,12 @@ public PlanWithProperties visitTableFinish(TableFinishNode node, PreferredProper
             return builder.build();
         }
 
-        private <T> Function<T, Optional<T>> createTranslator(SetMultimap<T, T> inputToOutput)
+        private Function<RowExpression, Optional<RowExpression>> createTranslator(SetMultimap<VariableReferenceExpression, VariableReferenceExpression> inputToOutput)
         {
-            return input -> inputToOutput.get(input).stream().findAny();
+            return input -> {
+                checkArgument(input instanceof VariableReferenceExpression, "Expect VariableReferenceExpression");
```

Comment:
nit: you can remove it. `(VariableReferenceExpression) input` will do the job and print the good error message